src/cmd/ksh93/sh/arith.c

```diff
@@ -295,71 +304,79 @@ static const Fltconst_t fltconst[] = {
-    if (nv_isattr(np, NV_DOUBLE) == NV_DOUBLE) {
-        const Fltconst_t *fp = fltconst;
-        n = sizeof(fltconst) / sizeof(Fltconst_t);
-        for (; n-- > 0; fp++) {
-            if (strcmp(fp->name, ep) == 0) break;
+    int unsign = nv_isattr(np, NV_LTOU);
+    node = *np;
+    node.nvalue.ldp = &dd;
+    if (nv_isattr(np, NV_SHORT)) {
+        if (unsign) {
+            node.nvalue.u = ip->us;
+        } else {
+            node.nvalue.i16 = ip->ss;
         }
-        if (n >= 0) {
-            node = *np;
-            node.nvalue.ldp = &dd;
-            if (nv_isattr(np, NV_SHORT)) {
-                *node.nvalue.fp = fp->f;
-            } else if (nv_isattr(np, NV_LONG)) {
-                *node.nvalue.ldp = fp->l;
-            } else {
-                *node.nvalue.dp = fp->d;
-            }
-            return &node;
+    } else if (nv_isattr(np, NV_LONG)) {
+        if (unsign) {
+            *node.nvalue.llp = ip->ul;
+        } else {
+            *node.nvalue.llp = ip->sl;
         }
     } else {
-        const Intconst_t *ip = intconst;
-        n = sizeof(intconst) / sizeof(Intconst_t);
-        for (; n-- >= 0; ip++) {
-            if (strcmp(ip->name, ep) == 0) break;
+        if (unsign) {
+            *node.nvalue.ip = ip->ui;
+        } else {
+            *node.nvalue.ip = ip->ul;
         }
-        if (n >= 0) {
-            int unsign = nv_isattr(np, NV_LTOU);
-            node = *np;
-            node.nvalue.ldp = &dd;
-            if (nv_isattr(np, NV_SHORT)) {
-                if (unsign) {
-                    node.nvalue.u = ip->us;
-                } else {
-                    node.nvalue.i16 = ip->ss;
-                }
-            } else if (nv_isattr(np, NV_LONG)) {
-                if (unsign) {
-                    *node.nvalue.llp = ip->ul;
-                } else {
-                    *node.nvalue.llp = ip->sl;
-                }
-            } else {
-                if (unsign) {
-                    *node.nvalue.ip = ip->ui;
-                } else {
-                    *node.nvalue.ip = ip->ul;
-                }
-            }
-            return &node;
+    }
+    return &node;
+}
+
+static_fn Namval_t *float_limit(const Fltconst_t *fp, Namval_t *np) {
+    static Namval_t node;
+    static Sfdouble_t dd;
+
+    node = *np;
+    node.nvalue.ldp = &dd;
+    if (nv_isattr(np, NV_SHORT)) {
+        *node.nvalue.fp = fp->f;
+    } else if (nv_isattr(np, NV_LONG)) {
+        *node.nvalue.ldp = fp->l;
+    } else {
+        *node.nvalue.dp = fp->d;
+    }
+    return &node;
+}
+
+static_fn Namval_t *check_limits(Shell_t *shp, char *cp) {
+    char *ep = strrchr(cp, '.');
+    if (!ep) return NULL;  // no attribute name
+
+    *ep = 0;
+    Namval_t *np = nv_open(cp, shp->var_tree, NV_VARNAME | NV_NOADD | NV_NOFAIL);
+    *ep++ = '.';
+    if (!np || !nv_isattr(np, NV_INTEGER)) return NULL;  // var type not integer or float
+
+    if (nv_isattr(np, NV_DOUBLE) == NV_DOUBLE) {
```

Comment:
IIRC you questioned this pattern, why are you using it ?

Reply:
Yes I did. See issue #855. But changing it is outside the scope of this fix. And omitting the `== NV_DOUBLE` is only safe if you know the symbol has a single bit set.

Reply:
Which is precisely why I want to change the definition of `nv_isattr()`. Notice that the preceding  use omits the `== NV_INTEGER` test. The problem here is that this is the definition for `NV_DOUBLE`:

```
#define NV_DOUBLE (NV_INTEGER | NV_ZFILL)
```

This is without a doubt one of the worst APIs I've seen I've seen in this project.